plugins/devtools-backend/src/util/Lockfile.ts

```diff
@@ -0,0 +1,317 @@
+      }
+
+      // Find all versions currently in use
+      const versions = Array.from(new Set(entries.map(e => e.version))).sort(
+        (v1, v2) => semver.rcompare(v1, v2),
+      );
+
+      // If we're not using at least 2 different versions we're done
+      if (versions.length < 2) {
+        continue;
+      }
+
+      const acceptedVersions = new Set<string>();
+      for (const { version, range } of entries) {
+        // Finds the highest matching version from the the known versions
+        // TODO(Rugvip): We may want to select the version that satisfies the most ranges rather than the highest one
+        const acceptedVersion = versions.find(v => semver.satisfies(v, range));
+        if (!acceptedVersion) {
+          throw new Error(
+            `No existing version was accepted for range ${range}, searching through ${versions}, for package ${name}`,
+          );
+        }
+
+        if (acceptedVersion !== version) {
+          result.newVersions.push({
+            name,
+            range,
+            newVersion: acceptedVersion,
+            oldVersion: version,
+          });
+        }
+
+        acceptedVersions.add(acceptedVersion);
+      }
+
+      // If all ranges were able to accept the same version, we're done
+      if (acceptedVersions.size === 1) {
+        continue;
+      }
+
+      // Find the max version that we may want bump older packages to
+      const maxVersion = Array.from(acceptedVersions).sort(semver.rcompare)[0];
+      // Find all existing ranges that satisfy the new max version, and pick the one that
+      // results in the highest minimum allowed version, usually being the more specific one
+      const maxEntry = entries
+        .filter(e => semver.satisfies(maxVersion, e.range))
+        .map(e => ({ e, min: semver.minVersion(e.range) }))
+        .filter(p => p.min)
+        .sort((a, b) => semver.rcompare(a.min!, b.min!))[0]?.e;
+      if (!maxEntry) {
+        throw new Error(
+          `No entry found that satisfies max version '${maxVersion}'`,
+        );
+      }
+
+      // Find all entries that don't satisfy the max version
+      for (const { version, range } of entries) {
+        if (semver.satisfies(maxVersion, range)) {
+          continue;
+        }
+
+        result.newRanges.push({
+          name,
+          oldRange: range,
+          newRange: maxEntry.range,
+          oldVersion: version,
+          newVersion: maxVersion,
+        });
+      }
+    }
+
+    return result;
+  }
+
+  remove(name: string, range: string): boolean {
+    const query = `${name}@${range}`;
+    const existed = Boolean(this.data[query]);
+    delete this.data[query];
+
+    const newEntries = this.packages.get(name)?.filter(e => e.range !== range);
+    if (newEntries) {
+      this.packages.set(name, newEntries);
+    }
+
+    return existed;
+  }
+
+  /** Modifies the lockfile by bumping packages to the suggested versions */
+  replaceVersions(results: AnalyzeResultNewVersion[]) {
```

Comment:
Do we need this? I guess this is just a copy of the `Lockfile` stuff from the `cli` right? Maybe we just trim this down to what we need for reading and parsing?

Reply:
Yes, for sure, I've slimed this down to only the things I'm actually using. This makes much more sense now 👍